internal/services/executor/executor.go

```diff
@@ -1517,3 +1531,50 @@ func (e *Executor) Run(ctx context.Context) error {
 
 	return nil
 }
+
+func (e *Executor) tasksTimeoutCleanerLoop(ctx context.Context) {
+	for {
+		e.log.Debug().Msgf("tasksTimeoutCleaner")
+
+		e.tasksTimeoutCleaner(ctx)
+
+		sleepCh := time.NewTimer(2 * time.Second).C
+		select {
+		case <-ctx.Done():
+			return
+		case <-sleepCh:
+		}
+	}
+}
+
+func (e *Executor) tasksTimeoutCleaner(ctx context.Context) {
+	for _, rtID := range e.runningTasks.ids() {
+		if rt, ok := e.runningTasks.get(rtID); ok {
+			rt.Lock()
+			if rt.et.Status.Phase == types.ExecutorTaskPhaseRunning && rt.et.Spec.TaskTimeoutInterval != 0 && time.Since(*rt.et.Status.StartTime) >= rt.et.Spec.TaskTimeoutInterval {
+				rt.et.Status.Phase = types.ExecutorTaskPhaseTimedout
+				rt.et.Status.EndTime = util.TimeP(time.Now())
+				rt.et.Status.FailError = "timedout"
+
+				if rt.et.Status.SetupStep.Phase == types.ExecutorTaskPhaseRunning {
+					rt.et.Status.SetupStep.Phase = types.ExecutorTaskPhaseFailed
+					rt.et.Status.SetupStep.EndTime = util.TimeP(time.Now())
+				}
+				for _, s := range rt.et.Status.Steps {
+					if s.Phase == types.ExecutorTaskPhaseRunning {
+						s.Phase = types.ExecutorTaskPhaseFailed
+						s.EndTime = util.TimeP(time.Now())
+					}
+				}
+
+				if err := e.sendExecutorTaskStatus(ctx, rt.et); err != nil {
+					e.log.Err(err).Send()
+				}
+
+				rt.cancel()
+				e.runningTasks.delete(rtID)
```

Comment:
don't delete from running tasks, we must wait that the task execution is finished and be sure that sending the status was successful before removing it.